benchmark/performance_utils.py

```diff
@@ -1,164 +1,348 @@
+            ]
+            if invalid_metrics:
+                raise ValueError(
+                    f"Invalid metrics: {', '.join(invalid_metrics)} for operation: '{self.op_name}'"
+                )
+            unsatisfied_metrics = check_metric_dependencies(user_desired_metrics)
+            if unsatisfied_metrics:
+                raise ValueError(
+                    f"Unsatisfied metric dependencies: {', '.join(unsatisfied_metrics)}"
+                )
+
+        self.to_bench_metrics = user_desired_metrics or self.metrics
+
+    def set_dtypes(self, user_desired_dtypes: Optional[List[torch.dtype]]):
+        # Validate user-specified dtypes
+        if user_desired_dtypes and not all(
+            dtype in self.dtypes for dtype in user_desired_dtypes
+        ):
+            invalid_dtypes = [
+                dtype for dtype in user_desired_dtypes if dtype not in self.dtypes
+            ]
+            raise ValueError(
+                f"Given dtype(s) '{', '.join(str(dtype) for dtype in invalid_dtypes)}'"
+                f"can't be supported by this op '{self.op_name}'"
+            )
+        self.to_bench_dtypes = (
+            user_desired_dtypes if user_desired_dtypes else self.dtypes
+        )
+
+    def set_shapes(self, op_specified_shapes: Optional[List[Any]] = None):
+        self.shapes = (
+            op_specified_shapes if op_specified_shapes else self.DEFAULT_SHAPES
+        )
+
+    def record_shapes(self, *args, **kwargs):
+        def deep_parse(item):
+            if isinstance(item, torch.Tensor):
+                return item.size()
+            elif isinstance(item, (int, float, str, torch.dtype)):
+                return item
+            elif isinstance(item, (list, tuple)):
+                return [deep_parse(sub_item) for sub_item in item]
+            elif isinstance(item, dict):
+                return {key: deep_parse(value) for key, value in item.items()}
+            return None
+
+        parsed_args = [deep_parse(arg) for arg in args]
+        parsed_kwargs = {key: deep_parse(value) for key, value in kwargs.items()}
+        if parsed_args and parsed_kwargs:
+            return parsed_args, parsed_kwargs
+        return parsed_args if parsed_args else parsed_kwargs
+
+    def init_user_config(self):
+        # TODO: device setting
+        self.cpu_mode = Config.cpu_mode
+        self.set_dtypes(Config.user_desired_dtypes)
+        self.set_metrics(Config.user_desired_metrics)
+        self.set_shapes()
 
     def set_gems(self, gems_op):
         self.gems_op = gems_op
 
-    def profile(self, op, *args, **kwargs):
+    def get_latency(self, op, *args, **kwargs):
         fn = lambda: op(*args, **kwargs)
         if self.is_backward:
             out = fn()
             dout = torch.randn_like(out)
             fn = lambda: out.backward(dout, retain_graph=True)
-        if CPU_MODE:
-            for i in range(WARMUP):
+        if Config.cpu_mode:
```

Comment:
could we design a new mode which outputs both cpu latency and gpu latency? @tianxiao-baai